numpy/lib/shape_base.py

```diff
@@ -1142,25 +1142,33 @@ def kron(a, b):
     b = asanyarray(b)
     a = array(a, copy=False, subok=True, ndmin=b.ndim)
     ndb, nda = b.ndim, a.ndim
+    nd = max(ndb, nda)
+
     if (nda == 0 or ndb == 0):
         return _nx.multiply(a, b)
+
     as_ = a.shape
     bs = b.shape
     if not a.flags.contiguous:
         a = reshape(a, as_)
     if not b.flags.contiguous:
         b = reshape(b, bs)
-    nd = ndb
-    if (ndb != nda):
-        if (ndb > nda):
-            as_ = (1,)*(ndb-nda) + as_
-        else:
-            bs = (1,)*(nda-ndb) + bs
-            nd = nda
-    result = outer(a, b).reshape(as_+bs)
-    axis = nd-1
-    for _ in range(nd):
-        result = concatenate(result, axis=axis)
+
+    # Equalise the shapes by prepending smaller one with 1s
+    as_ = (1,)*max(0, ndb-nda) + as_
+    bs = (1,)*max(0, nda-ndb) + bs
+
+    # Compute the product
+    a_arr = _nx.asarray(a).reshape(a.size, 1)
```

Comment:
This broke astropy (see https://github.com/astropy/astropy/issues/13023), which is counting on subclasses to be treated correctly. Since `a` and `b` are already arrays, why is this done? If it is for matrices, it would be better to just explicitly test for those.

Reply:
Hi @mhvk. Really sorry for the regression. I have already raised a PR: #21262. This is fixing the issue:
```python
import numpy as np
print(np.__version__)
np.__version__ = '1.23' # aviod asscalar deprecation error

import astropy
from astropy.utils.masked import Masked
print(astropy.__version__)

a = np.arange(6.).reshape(2, 3)
mask_a = np.array([[True, False, False],[False, True, False]])
b = np.array([-3., -2., -1.])
mask_b = np.array([False, True, False])

ma = Masked(a, mask=mask_a)
mb = Masked(b, mask=mask_b)

print(np.kron(ma, mb))
```

Output:
```python
1.22.0.dev0+2848.g20dcd5810
5.1.dev767+g323dd7fe2
[[ ———  ———  ———  -3.  ———  -1.  -6.  ———  -2.]
 [ -9.  ———  -3.  ———  ———  ——— -15.  ———  -5.]]

```

I'll try to get it in quickly today.